### DISCUSSION TITLE
EMQX SSL/TLS connection with the verify=verify_peer option

### DISCUSSION THREAD

**Discussion by anushageorgewa** [emqx/emqx#13681] [General]
Hai,
I have an Emqx Server (running as docker container). It has a valid Lertsencrpt certificates.
I have created the client certificate using the CA certificates.
May client needs to provide the client certificate during handshake. 
While connectiong the client gets the Error : ssl.SSLError: [SSL: TLSV1_ALERT_UNKNOWN_CA] tlsv1 alert unknown ca (_ssl.c:2633)

On the server i have mentioned it as:
listeners.ssl.default {
  bind = "0.0.0.0:8883"
  max_connections = 1024000
  ssl_options {
    cacertfile = "etc/letsencrypt/live/domain/cacert.pem"
    certfile = "etc/letsencrypt/live/domain/cert.pem"
    keyfile = "etc/letsencrypt/live/domain/key.pem"
    verify = verify_peer
    fail_if_no_peer_cert = true
  }
}

Can you help on this, why my clients are getting the Unknown CA issue eventhough the issued CA has mentioned?

**Comment by id** [comment]
@anushageorgewa you need to use cacert.pem from Letsencrypt on client side as well, so that it can validate server's certificate.

**Reply by anushageorgewa** [reply]
Thank you for the quick reply. On the client side i use the configuration, 

context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
        
  context.load_cert_chain(
      certfile="client.crt",
      keyfile="client.key",
      password=CERT_PASSWORD
  )
  context.load_verify_locations(cafile="cacert.pem")
  self.client.tls_set_context(context)


so that it should work I hope. Is there anything to do with this ?

**Reply by anushageorgewa** [reply]
This is the Exception which I gets, while connecting to the Server.
.................

    command = self._sock_recv(1)

  File "/root/.local/share/virtualenvs/modular-system-5Iu_2Dgr/lib/python3.9/site-packages/paho/mqtt/client.py", line 640, in _sock_recv

    return self._sock.recv(bufsize)

  File "/usr/local/lib/python3.9/ssl.py", line 1260, in recv

    return self.read(buflen)

  File "/usr/local/lib/python3.9/ssl.py", line 1135, in read

    return self._sslobj.read(len)

ssl.SSLError: [SSL: TLSV1_ALERT_UNKNOWN_CA] tlsv1 alert unknown ca (_ssl.c:2633)


On my emqxserver I have **cert.pem  chain.pem  fullchain.pem  privkey.pem**. I used the cert.pem and the privkey.pem to create my client certificates. The fullchain.pem is actually my cacert.pem I mentioned above. 
So I don't understand why the client is getting this error. Using openssl verify command with the cert.pem and client key it says OK on the Server. 
Can you help on this ?

**Reply by id** [reply]
How exactly did you create client certificate?

**Reply by anushageorgewa** [reply]
1. openssl genrsa -out client.key 2048
2. openssl req -new -key client.key -out client.csr -subj "/CN=client"
3. openssl x509 -req -in client.csr -CA /etc/letsencrypt/live/domain/cert.pem -CAkey /etc/letsencrypt/live/domain/privkey.pem -CAcreateserial -out client.crt -days 365 -sha256

This is the steps which I have used. As a result I got client.crt client.csr and client.key

**Reply by id** [reply]
You need to sign the cert with a CA and CA key, not with server certificate.

Server need to trust CA which signed client certificate, and client need to trust CA which signed server certificate.

**Reply by anushageorgewa** [reply]
On the third command I specify to use -CA and CAKey to sign the client certificate. And I copy the client.cert, client.key and cert.pem to the client side , so that it is trusted right?

Should I make any modifications to the commands ? Could you please mention how to generate it properly ?
thank you in advance.

**Reply by id** [reply]
It's not about the command, you can't use /etc/letsencrypt/live/domain/cert.pem file to sign client certificate. You need a certificate authority.

You can use my scripts to generate a self-signed CA and a client certificate: https://github.com/id/mkcerts.
```
./ca.sh
./client.sh
```

Then you need add `MyCA/ca.pem` to `cacertfile` field for EMQX SSL listener, and use `client.pem` and `client.key` for client connections.

**Reply by anushageorgewa** [reply]
Wow, this has worked. 
Thank you so much for your quick Support !
Have a nice Weekend.
